File: interview/sample/problems/WordSearch.java
public class WordSearch {
    public boolean exist(char[][] board, String word) {
        int rows = board.length;
        int cols = board[0].length;

        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                if (board[row][col] == word.charAt(0) && dfs(board, row, col, 0, word)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean dfs(char[][] board, int row, int col, int index, String word) {
        // If the full word is matched
        if (index == word.length()) {
            return true;
        }

        // Check boundaries and if the cell matches the current character
        if (row < 0 || row >= board.length || col < 0 || col >= board[0].length || board[row][col] != word.charAt(index)) {
            return false;
        }

        // Mark the cell as visited (use a placeholder)
        char temp = board[row][col];
        board[row][col] = '#';

        // Explore all four directions
        boolean found = dfs(board, row + 1, col, index + 1, word)
                || dfs(board, row - 1, col, index + 1, word)
                || dfs(board, row, col + 1, index + 1, word)
                || dfs(board, row, col - 1, index + 1, word);

        // Backtrack by restoring the cell
        board[row][col] = temp;
        return found;
    }

    public static void main(String[] args) {
        WordSearch ws = new WordSearch();
        char[][] board = {
            {'A', 'B', 'C', 'E'},
            {'S', 'F', 'C', 'S'},
            {'A', 'D', 'E', 'E'}
        };
        String word = "ABCCED";
        System.out.println(ws.exist(board, word));  // Output: true
    }
}

File: interview/sample/problems/word_search.py
def exist(board, word):
    rows, cols = len(board), len(board[0])

    def dfs(row, col, index):
        # If the full word is matched
        if index == len(word):
            return True
        
        # Check boundaries and if the cell matches the current character
        if row < 0 or row >= rows or col < 0 or col >= cols or board[row][col] != word[index]:
            return False
        
        # Mark the cell as visited (use a placeholder)
        temp = board[row][col]
        board[row][col] = '#'
        
        # Explore all four directions
        found = (
            dfs(row + 1, col, index + 1) or
            dfs(row - 1, col, index + 1) or
            dfs(row, col + 1, index + 1) or
            dfs(row, col - 1, index + 1)
        )
        
        # Backtrack by restoring the cell
        board[row][col] = temp
        return found

    # Start DFS from each cell
    for r in range(rows):
        for c in range(cols):
            if board[r][c] == word[0] and dfs(r, c, 0):
                return True
    
    return False

# Example Usage
board = [
    ['A', 'B', 'C', 'E'],
    ['S', 'F', 'C', 'S'],
    ['A', 'D', 'E', 'E']
]
word = "ABCCED"
print(exist(board, word))  # Output: True
